### ISSUE TITLE
hotplug enumerate regression in 1.0.25

### ISSUE THREAD

**Issue by lunixbochs** [libusb/libusb#1082]
With roughly this code, the enumerated device list is sent twice to the hotplug callback on initialization.
I bisected the problem to https://github.com/libusb/libusb/commit/6929b8270170a76e9af30d0a9e95c2081f4372b0 

Tested on macOS 12, M1, with libusb 1.0.25.
In 1.0.24 and prior, the list of devices was correctly sent only one time to the callback.

```
static int hotplug_event(
        struct libusb_context *ctx,
        struct libusb_device *dev,
        libusb_hotplug_event event,
        void *user) {
    ...
}

    libusb_hotplug_callback_handle handle = 0;
    int rc = libusb_hotplug_register_callback(eye_uctx.ctx,
        LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED | LIBUSB_HOTPLUG_EVENT_DEVICE_LEFT,
        LIBUSB_HOTPLUG_ENUMERATE,
        LIBUSB_HOTPLUG_MATCH_ANY, LIBUSB_HOTPLUG_MATCH_ANY, LIBUSB_HOTPLUG_MATCH_ANY,
        hotplug_event, cb, &handle);
```

**Comment by mcuee** [comment]
Please help to post the debug log by setting LIBUSB_DEBUG=4 and then run your application. Thanks. Does this happen on other OS like Windows or Linux?

**Comment by gsurkov** [comment]
I can confirm the same issue on linux (Arch Linux, 5.16.10, libusb 1.0.25).
The code I am using to register the hotplug:
```c++
const auto events = libusb_hotplug_event(LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED | LIBUSB_HOTPLUG_EVENT_DEVICE_LEFT);
const auto err = libusb_hotplug_register_callback(nullptr, events, LIBUSB_HOTPLUG_ENUMERATE, info.vendorID(),
                          info.productID(), LIBUSB_HOTPLUG_MATCH_ANY, libusbHotplugCallback, this, nullptr);
```
If I remove the `LIBUSB_HOTPLUG_ENUMERATE` flag, the callback gets activated only once, but then it stops working on the earlier versions.
Version 1.0.24 works just fine, only one enumerate event if the device was already connected and only one arrived event if it was connected during the application run time.

Below is the output with `LIBUSB_DEBUG=4` and libusb = 1.0.25.
```
[timestamp] [threadID] facility level [function call] <message>
--------------------------------------------------------------------------------
[ 0.000021] [0002d815] libusb: debug [libusb_init] created default context
[ 0.000031] [0002d815] libusb: debug [libusb_init] libusb v1.0.25.11692
[ 0.000041] [0002d815] libusb: debug [usbi_add_event_source] add fd 13 events 1
[ 0.000051] [0002d815] libusb: debug [usbi_io_init] using timer for timeouts
[ 0.000057] [0002d815] libusb: debug [usbi_add_event_source] add fd 14 events 1
[ 0.000068] [0002d815] libusb: debug [get_kernel_version] reported kernel version is 5.16.10-arch1-1
[ 0.000087] [0002d815] libusb: debug [op_init] found usbfs at /dev/bus/usb
[ 0.000109] [0002d815] libusb: debug [op_init] max iso packet length is (likely) 98304 bytes
[ 0.000128] [0002d815] libusb: debug [op_init] sysfs is available
[ 0.000397] [0002d81b] libusb: debug [linux_udev_event_thread_main] udev event thread entering
[ 0.007806] [0002d815] libusb: debug [linux_get_device_address] getting address for device: usb1 detached: 0
[ 0.007817] [0002d815] libusb: debug [linux_get_device_address] scan usb1
[ 0.007851] [0002d815] libusb: debug [linux_get_device_address] bus=1 dev=1
[ 0.007856] [0002d815] libusb: debug [linux_enumerate_device] busnum 1 devaddr 1 session_id 257
[ 0.007860] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 1/1 (session 257)
[ 0.007938] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 1-1 detached: 0
[ 0.007943] [0002d815] libusb: debug [linux_get_device_address] scan 1-1
[ 0.007968] [0002d815] libusb: debug [linux_get_device_address] bus=1 dev=2
[ 0.007973] [0002d815] libusb: debug [linux_enumerate_device] busnum 1 devaddr 2 session_id 258
[ 0.007978] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 1/2 (session 258)
[ 0.008006] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dbf2f90 (1-1) has parent 0x55bc3dc2de80 (usb1) port 1
[ 0.008083] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 1-1.2 detached: 0
[ 0.008089] [0002d815] libusb: debug [linux_get_device_address] scan 1-1.2
[ 0.008116] [0002d815] libusb: debug [linux_get_device_address] bus=1 dev=17
[ 0.008121] [0002d815] libusb: debug [linux_enumerate_device] busnum 1 devaddr 17 session_id 273
[ 0.008126] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 1/17 (session 273)
[ 0.008152] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dc2dd40 (1-1.2) has parent 0x55bc3dbf2f90 (1-1) port 2
[ 0.008219] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 1-1.4 detached: 0
[ 0.008224] [0002d815] libusb: debug [linux_get_device_address] scan 1-1.4
[ 0.008254] [0002d815] libusb: debug [linux_get_device_address] bus=1 dev=82
[ 0.008259] [0002d815] libusb: debug [linux_enumerate_device] busnum 1 devaddr 82 session_id 338
[ 0.008262] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 1/82 (session 338)
[ 0.008289] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dbf36c0 (1-1.4) has parent 0x55bc3dbf2f90 (1-1) port 4
[ 0.008357] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 1-1.6 detached: 0
[ 0.008362] [0002d815] libusb: debug [linux_get_device_address] scan 1-1.6
[ 0.008388] [0002d815] libusb: debug [linux_get_device_address] bus=1 dev=5
[ 0.008392] [0002d815] libusb: debug [linux_enumerate_device] busnum 1 devaddr 5 session_id 261
[ 0.008396] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 1/5 (session 261)
[ 0.008421] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dbf30d0 (1-1.6) has parent 0x55bc3dbf2f90 (1-1) port 6
[ 0.008482] [0002d815] libusb: debug [linux_get_device_address] getting address for device: usb3 detached: 0
[ 0.008487] [0002d815] libusb: debug [linux_get_device_address] scan usb3
[ 0.008512] [0002d815] libusb: debug [linux_get_device_address] bus=3 dev=1
[ 0.008516] [0002d815] libusb: debug [linux_enumerate_device] busnum 3 devaddr 1 session_id 769
[ 0.008521] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 3/1 (session 769)
[ 0.008608] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 3-1 detached: 0
[ 0.008614] [0002d815] libusb: debug [linux_get_device_address] scan 3-1
[ 0.008640] [0002d815] libusb: debug [linux_get_device_address] bus=3 dev=2
[ 0.008644] [0002d815] libusb: debug [linux_enumerate_device] busnum 3 devaddr 2 session_id 770
[ 0.008647] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 3/2 (session 770)
[ 0.008669] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dbf4030 (3-1) has parent 0x55bc3dbf2b70 (usb3) port 1
[ 0.008733] [0002d815] libusb: debug [linux_get_device_address] getting address for device: usb4 detached: 0
[ 0.008738] [0002d815] libusb: debug [linux_get_device_address] scan usb4
[ 0.008760] [0002d815] libusb: debug [linux_get_device_address] bus=4 dev=1
[ 0.008764] [0002d815] libusb: debug [linux_enumerate_device] busnum 4 devaddr 1 session_id 1025
[ 0.008778] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 4/1 (session 1025)
[ 0.008854] [0002d815] libusb: debug [linux_get_device_address] getting address for device: usb2 detached: 0
[ 0.008860] [0002d815] libusb: debug [linux_get_device_address] scan usb2
[ 0.008881] [0002d815] libusb: debug [linux_get_device_address] bus=2 dev=1
[ 0.008884] [0002d815] libusb: debug [linux_enumerate_device] busnum 2 devaddr 1 session_id 513
[ 0.008888] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 2/1 (session 513)
[ 0.008962] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 2-1 detached: 0
[ 0.008967] [0002d815] libusb: debug [linux_get_device_address] scan 2-1
[ 0.008994] [0002d815] libusb: debug [linux_get_device_address] bus=2 dev=2
[ 0.008998] [0002d815] libusb: debug [linux_enumerate_device] busnum 2 devaddr 2 session_id 514
[ 0.009002] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 2/2 (session 514)
[ 0.009029] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dbf3fa0 (2-1) has parent 0x55bc3dbf35c0 (usb2) port 1
[ 0.009096] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 2-1.2 detached: 0
[ 0.009102] [0002d815] libusb: debug [linux_get_device_address] scan 2-1.2
[ 0.009128] [0002d815] libusb: debug [linux_get_device_address] bus=2 dev=3
[ 0.009133] [0002d815] libusb: debug [linux_enumerate_device] busnum 2 devaddr 3 session_id 515
[ 0.009138] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 2/3 (session 515)
[ 0.009164] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dbf44b0 (2-1.2) has parent 0x55bc3dbf3fa0 (2-1) port 2
[ 0.009232] [0002d815] libusb: debug [linux_get_device_address] getting address for device: 2-1.4 detached: 0
[ 0.009238] [0002d815] libusb: debug [linux_get_device_address] scan 2-1.4
[ 0.009263] [0002d815] libusb: debug [linux_get_device_address] bus=2 dev=10
[ 0.009268] [0002d815] libusb: debug [linux_enumerate_device] busnum 2 devaddr 10 session_id 522
[ 0.009271] [0002d815] libusb: debug [linux_enumerate_device] allocating new device for 2/10 (session 522)
[ 0.009296] [0002d815] libusb: debug [linux_get_parent_info] dev 0x55bc3dbf31d0 (2-1.4) has parent 0x55bc3dbf3fa0 (2-1) port 4
[ 0.009383] [0002d815] libusb: debug [libusb_hotplug_register_callback] new hotplug cb 0x55bc3dbe2680 with handle 1
[ 0.009388] [0002d815] libusb: debug [libusb_get_device_list]  
[ 0.009397] [0002d815] libusb: debug [libusb_hotplug_register_callback] new hotplug cb 0x55bc3dc2b0c0 with handle 2
[ 0.009402] [0002d815] libusb: debug [libusb_get_device_list]  
[ 0.009411] [0002d815] libusb: debug [libusb_get_device_descriptor]  
[ 0.381305] [0002d815] libusb: debug [libusb_get_device_descriptor]  
[ 0.381352] [0002d815] libusb: debug [libusb_open] open 1.17
[ 0.509157] [0002d815] libusb: debug [usbi_add_event_source] add fd 33 events 4
[ 0.509185] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fa40c78
[ 0.509190] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.509206] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.509211] [0002d815] libusb: debug [handle_events] event sources modified, reallocating event data
[ 0.509216] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.509222] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.509227] [0002d815] libusb: debug [handle_event_trigger] event triggered
[ 0.509231] [0002d815] libusb: debug [handle_event_trigger] hotplug message received
[ 0.509238] [0002d815] libusb: debug [libusb_get_device_descriptor]  
[ 0.509253] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.509258] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.509609] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.509620] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=4
[ 0.509626] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.509632] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.509639] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fa40c78 has callback 0x7f2d97b3c540
[ 0.509645] [0002d815] libusb: debug [sync_transfer_cb] actual_length=4
[ 0.509651] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fa40c78
[ 0.509657] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fa40d98
[ 0.509661] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.509672] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.509677] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.510086] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.510163] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=42
[ 0.510179] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.510193] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.510208] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fa40d98 has callback 0x7f2d97b3c540
[ 0.510222] [0002d815] libusb: debug [sync_transfer_cb] actual_length=42
[ 0.510237] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fa40d98
[ 0.510260] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fa40e78
[ 0.510300] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.510321] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.510336] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.510552] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.510560] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=4
[ 0.510566] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.510571] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.510576] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fa40e78 has callback 0x7f2d97b3c540
[ 0.510582] [0002d815] libusb: debug [sync_transfer_cb] actual_length=4
[ 0.510588] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fa40e78
[ 0.510592] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fa40f58
[ 0.510597] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.510608] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.510613] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.511023] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.511034] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=32
[ 0.511041] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.511047] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.511053] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fa40f58 has callback 0x7f2d97b3c540
[ 0.511058] [0002d815] libusb: debug [sync_transfer_cb] actual_length=32
[ 0.511064] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fa40f58
[ 0.511072] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fa41088
[ 0.511076] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.511086] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.511091] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.511288] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.511294] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=4
[ 0.511299] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.511304] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.511310] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fa41088 has callback 0x7f2d97b3c540
[ 0.511315] [0002d815] libusb: debug [sync_transfer_cb] actual_length=4
[ 0.511321] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fa41088
[ 0.511326] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fa41168
[ 0.511330] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.511345] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.511351] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.511758] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.511767] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=26
[ 0.511772] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.511778] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.511784] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fa41168 has callback 0x7f2d97b3c540
[ 0.511789] [0002d815] libusb: debug [sync_transfer_cb] actual_length=26
[ 0.511795] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fa41168
[ 0.511803] [0002d815] libusb: debug [libusb_close]  
[ 0.511809] [0002d815] libusb: debug [usbi_remove_event_source] remove fd 33
[DEVREG] Detected new device: VID_0x483:PID_0x5740
[ 0.513338] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.513351] [0002d815] libusb: debug [handle_events] event sources modified, reallocating event data
[ 0.513373] [0002d815] libusb: debug [usbi_wait_for_events] poll() 2 fds with timeout in 10ms
[ 0.523452] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 0
[ 0.576357] [0002d815] libusb: debug [libusb_get_device_descriptor]  
[ 0.576376] [0002d815] libusb: debug [libusb_open] open 1.17
[ 0.576420] [0002d815] libusb: debug [usbi_add_event_source] add fd 34 events 4
[ 0.576433] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fb8d078
[ 0.576439] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.576463] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.576468] [0002d815] libusb: debug [handle_events] event sources modified, reallocating event data
[ 0.576475] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.576758] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.576775] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=4
[ 0.576779] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.576784] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.576789] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fb8d078 has callback 0x7f2d97b3c540
[ 0.576793] [0002d815] libusb: debug [sync_transfer_cb] actual_length=4
[ 0.576799] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fb8d078
[ 0.576805] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fb8d078
[ 0.576808] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.576819] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.576823] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.577190] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.577218] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=42
[ 0.577231] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.577244] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.577257] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fb8d078 has callback 0x7f2d97b3c540
[ 0.577271] [0002d815] libusb: debug [sync_transfer_cb] actual_length=42
[ 0.577284] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fb8d078
[ 0.577302] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fb8d078
[ 0.577315] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.577338] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.577352] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.577657] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.577672] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=4
[ 0.577678] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.577684] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.577692] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fb8d078 has callback 0x7f2d97b3c540
[ 0.577698] [0002d815] libusb: debug [sync_transfer_cb] actual_length=4
[ 0.577705] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fb8d078
[ 0.577712] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fb8d078
[ 0.577717] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.577732] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.577738] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.578131] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.578166] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=32
[ 0.578179] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.578192] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.578206] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fb8d078 has callback 0x7f2d97b3c540
[ 0.578218] [0002d815] libusb: debug [sync_transfer_cb] actual_length=32
[ 0.578232] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fb8d078
[ 0.578248] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fb8d078
[ 0.578266] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.578300] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.578314] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.578606] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.578628] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=4
[ 0.578641] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.578653] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.578666] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fb8d078 has callback 0x7f2d97b3c540
[ 0.578679] [0002d815] libusb: debug [sync_transfer_cb] actual_length=4
[ 0.578693] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fb8d078
[ 0.578706] [0002d815] libusb: debug [libusb_submit_transfer] transfer 0x55bc3fb8d078
[ 0.578718] [0002d815] libusb: debug [add_to_flying_list] arm timer for timeout in 1000ms (first in line)
[ 0.578738] [0002d815] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.578751] [0002d815] libusb: debug [usbi_wait_for_events] poll() 3 fds with timeout in 60000ms
[ 0.579073] [0002d815] libusb: debug [usbi_wait_for_events] poll() returned 1
[ 0.579088] [0002d815] libusb: debug [reap_for_handle] urb type=2 status=0 transferred=26
[ 0.579093] [0002d815] libusb: debug [handle_control_completion] handling completion status 0
[ 0.579096] [0002d815] libusb: debug [arm_timer_for_next_timeout] no timeouts, disarming timer
[ 0.579102] [0002d815] libusb: debug [usbi_handle_transfer_completion] transfer 0x55bc3fb8d078 has callback 0x7f2d97b3c540
[ 0.579108] [0002d815] libusb: debug [sync_transfer_cb] actual_length=26
[ 0.579113] [0002d815] libusb: debug [libusb_free_transfer] transfer 0x55bc3fb8d078
[ 0.579120] [0002d815] libusb: debug [libusb_close]  
[ 0.579127] [0002d815] libusb: debug [usbi_remove_event_source] remove fd 34
[DEVREG] Detected new device: VID_0x483:PID_0x5740
```

**Comment by benzea** [comment]
Rather than doing the change from  https://github.com/libusb/libusb/commit/6929b8270170a76e9af30d0a9e95c2081f4372b0, one could also guard `usbi_hotplug_exit` itself by testing whether `hotplug_ready` is set. Then calling `usbi_hotplug_exit` unconditionally in `libusb_init` is fine.

----

Note, that applications are (currently) required to handle this corner case. libusb tried to suppress the events for the initial enumeration. However, there is *no* guarantee that the application will not receive two notifications anyway. There are race conditions for hotplug events occurring between libusb_init and registering the callback.

Quoting the documentation:

> ... This means that when using [LIBUSB_HOTPLUG_ENUMERATE](https://libusb.sourceforge.io/api-1.0/group__libusb__hotplug.html#gga95ce2ec85dec177854c423cea2853c52a01a1c0c9f4e6852cb6df713f2e8a7f6b), your callback may be called twice for the arrival of the same device, once from [libusb_hotplug_register_callback()](https://libusb.sourceforge.io/api-1.0/group__libusb__hotplug.html#ga5ab3955e2110a3099497a66256fb7fab) and once from [libusb_handle_events()](https://libusb.sourceforge.io/api-1.0/group__libusb__poll.html#ga4989086e3f0327f3886a4c474ec7c327); and/or your callback may be called for the removal of a device for which an arrived call was never made.

To avoid that, libusb would need to track the events on a per callback level.

**Comment by benzea** [comment]
OK, pushed the proposed patch to fix this into #1078 and included a regression test for this specific issue. No regression test was added for #989 as it is impossible to inject an error into the right location using umockdev.
